specification/servicefabric/resource-manager/Microsoft.ServiceFabric/preview/2020-12-01-preview/cluster.json

```diff
@@ -0,0 +1,1734 @@
+        ],
+        "responses": {
+          "200": {
+            "description": "The operation completed successfully.",
+            "schema": {
+              "$ref": "#/definitions/ClusterCodeVersionsListResult"
+            }
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.ServiceFabric/locations/{location}/clusterVersions": {
+      "get": {
+        "operationId": "ClusterVersions_List",
+        "summary": "Gets the list of Service Fabric cluster code versions available for the specified location.",
+        "description": "Gets all available code versions for Service Fabric cluster resources by location.",
+        "parameters": [
+          {
+            "$ref": "#/parameters/locationForClusterCodeVersions"
+          },
+          {
+            "$ref": "#/parameters/api-version"
+          },
+          {
+            "$ref": "#/parameters/subscriptionId"
+          }
+        ],
+        "x-ms-examples": {
+          "List cluster versions": {
+            "$ref": "./examples/ClusterVersionsList_example.json"
+          }
+        },
+        "tags": [
+          "ClusterVersion"
+        ],
+        "responses": {
+          "200": {
+            "description": "The operation completed successfully.",
+            "schema": {
+              "$ref": "#/definitions/ClusterCodeVersionsListResult"
+            }
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.ServiceFabric/locations/{location}/environments/{environment}/clusterVersions": {
+      "get": {
+        "operationId": "ClusterVersions_ListByEnvironment",
+        "summary": "Gets the list of Service Fabric cluster code versions available for the specified environment.",
+        "description": "Gets all available code versions for Service Fabric cluster resources by environment.",
+        "parameters": [
+          {
+            "$ref": "#/parameters/locationForClusterCodeVersions"
+          },
+          {
+            "$ref": "#/parameters/environment"
+          },
+          {
+            "$ref": "#/parameters/api-version"
+          },
+          {
+            "$ref": "#/parameters/subscriptionId"
+          }
+        ],
+        "x-ms-examples": {
+          "List cluster versions by environment": {
+            "$ref": "./examples/ClusterVersionsListByEnvironment.json"
+          }
+        },
+        "tags": [
+          "ClusterVersion"
+        ],
+        "responses": {
+          "200": {
+            "description": "The operation completed successfully.",
+            "schema": {
+              "$ref": "#/definitions/ClusterCodeVersionsListResult"
+            }
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ServiceFabric/clusters/{clusterName}/upgradableVersion/maximum": {
+      "post": {
+        "operationId": "UpgradableVersionMaximum_Post",
```

Comment:
Separating this out into 3 POST actions is excessive (and leads to the path problems you are running into).  Can you re-model this as a single "/clusters/{clusterName}/listUpgradableVersions" POST action that takes an optional request body.  If the body is not provided it returns both the min and max version in the payload.  If a targetVersion is provided in the body then it returns min, max, and the path to the target.  Using the "list" prefix will also allow users to retrieve this information in a template deployment (which seems valuable)

Reply:
@pilor thanks for the suggestion. We have changed it to have a single post action listUpgradableVersions

Reply:
@pilor I made the changes suggested. Can we get this pr merged?